harness/determined/profiler.py

```diff
@@ -0,0 +1,559 @@
+            return TrialProfilerMetricsBatch(values, batches, timestamps, labels)
+
+        def make_labels(name: str, metric_type: str, gpu_uuid_label: str = "") -> Dict[str, Any]:
+            return {
+                "trialId": self.trial_id,
+                "name": name,
+                "agentId": self.agent_id,
+                "gpuUuid": gpu_uuid_label,
+                "metricType": metric_type,
+            }
+
+        trial_profiler_metrics_batches = []
+        for metric_name in self.batch.keys():
+            # TODO: Don't forget to include GPU Memory
+            if metric_name != SysMetricType.GPU_UTIL_METRIC and len(self.batch[metric_name]) > 0:
+                trial_profiler_metrics_batches.append(
+                    to_post_format(
+                        self.batch[metric_name],
+                        make_labels(metric_name, SYSTEM_METRIC_TYPE_ENUM),
+                    )
+                )
+
+            # GPU Metrics need to be grouped by GPU UUID
+            # TODO: Don't forget to include GPU Memory
+            if (
+                metric_name == SysMetricType.GPU_UTIL_METRIC
+                and len(self.batch[metric_name].keys()) > 0
+            ):
+                for gpu_uuid in self.batch[metric_name].keys():
+                    trial_profiler_metrics_batches.append(
+                        to_post_format(
+                            self.batch[metric_name][gpu_uuid],
+                            make_labels(
+                                metric_name, SYSTEM_METRIC_TYPE_ENUM, gpu_uuid_label=gpu_uuid
+                            ),
+                        )
+                    )
+
+        return trial_profiler_metrics_batches
+
+
+class StartMessage:
+    pass
+
+
+class ShutdownMessage:
+    pass
+
+
+class SysMetricCollectorThread(threading.Thread):
+    """
+    Background thread for collecting profiler metrics at a high granularity and shipping them to
+    the master
+
+    - SimpleCpuUtilization = Measured in percent
+    - FreeMemory = Measured in Gigabytes
+    - NetworkSentThroughput = Measured in Gigabit/s
+    - NetworkRecvThroughput = Measured in Gigabit/s
+    - DiskIops
+    - DiskReadThroughput = Measured in bytes/second
+    - DiskWriteThroughput = Measured in bytes/second
+    - GpuUtilization = Measured in percent
+    """
+
+    FLUSH_INTERVAL = 10  # How often to make API calls
+    MEASUREMENT_INTERVAL = 0.1
+
+    def __init__(self, trial_id: int, agent_id: str, send_queue: queue.Queue):
+
+        self.current_batch_idx = 0
+        self.send_queue = send_queue
+        self.control_queue: "queue.Queue[Union['StartMessage', 'ShutdownMessage']]" = queue.Queue()
+        self.current_batch = SysMetricBatcher(trial_id, agent_id)
+        self.current_batch.clear()
+
+        super().__init__(daemon=True)
+
+    def activate(self) -> None:
+        self.control_queue.put(StartMessage())
+
+    def kill(self) -> None:
```

Comment:
This is actually "stop" rather than "kill".

```suggestion
    def stop(self) -> None:
```

Reply:
Why is this "stop" rather than "kill"? It doesn't complete active work before shutting down, it just drops any in-progress batches, which seems more like "kill" to me.